src/foam/java/refinements.js

```diff
@@ -2123,35 +2123,69 @@ foam.CLASS({
     function buildJavaClass(cls) {
       this.SUPER(cls);
       var capitalizedName = foam.String.capitalize(this.name);
+      var constantizedRefName = foam.String.constantize(this.referencedProperty.name);
 
-      cls.field({
-        name: `cached${capitalizedName}_`,
-        visibility: 'private',
-        type: `java.lang.ref.SoftReference<${this.of.id}>`
-      });
+      if ( this.cache ) {
+        cls.field({
+          name: `cached${capitalizedName}_`,
+          visibility: 'private',
+          type: `java.lang.ref.SoftReference<${this.of.id}>`
+        });
+        cls.field({
+          name: `${capitalizedName}CachedFor_`,
+          visibility: 'private',
+          type: 'Long'
+        });
+      }
 
-      cls.method({
-        name: `find${foam.String.capitalize(this.name)}`,
-        visibility: 'public',
-        type: this.of.id,
-        args: [ { name: 'x', type: 'foam.lang.X' } ],
-        body: `
+      var body = this.cache ? `
+          var referencedDAO = (foam.dao.DAO) x.get("${this.unauthorizedTargetDAOKey || this.targetDAOKey}");
           synchronized(this) {
+            var subject = (foam.core.auth.Subject) x.get("subject");
+            var user = subject == null ? null : subject.getUser();
             if ( this.cached${capitalizedName}_ != null && this.cached${capitalizedName}_.get() != null ) {
               var referent = this.cached${capitalizedName}_.get();
               var refPropVal = get${capitalizedName}();
 
-              if ( foam.util.SafetyUtil.equals(referent.getProperty("${this.referencedProperty.name}"), refPropVal) )
+              if ( user.getId() == ${capitalizedName}CachedFor_ && 
+                foam.util.SafetyUtil.equals(referent.getProperty("${this.referencedProperty.name}"), refPropVal) )
                 return this.cached${capitalizedName}_.get();
             }
 
-            var res = (${this.of.id})((foam.dao.DAO) x.get("${this.unauthorizedTargetDAOKey || this.targetDAOKey}")).find_(x, (Object) get${capitalizedName}());
-            if ( res != null )
+            var res = (${this.of.id}) referencedDAO.find_(x, (Object) get${capitalizedName}());
+            if ( res != null && user != null ) {
```

Comment:
Should we also not cache for 'system' user?   Makes no difference to test results - all still pass with your other changes. 
```suggestion
            if ( res != null && user != null && user.getId() == 1 ) {
```

Reply:
shouldnt it be id ！= 1 ？ you mean we shouldnt cache for system user right？